core/Pos/Core/Configuration.hs

```diff
@@ -0,0 +1,442 @@
+import qualified Data.ByteString            as BS
+import qualified Data.ByteString.Lazy       as BSL
+import           Data.Reflection            (Given (..), give)
+import           Data.Time.Units            (Microsecond, Millisecond, Second, convertUnit)
+import           Formatting                 (sformat, (%), shown)
+import           Text.JSON.Canonical        (FromJSON (..), parseCanonicalJSON)
+import           Serokell.Data.Memory.Units (Byte)
+import           System.Wlog                (WithLogger, logInfo)
+
+import           Pos.Core.Fee               (TxFeePolicy)
+import           Pos.Util.Util              ()
+import           Pos.Core.Constants         (genesisDataDigest)
+import           Pos.Core.Types             (BlockCount (..), SlotCount,
+                                             BlockVersionData (..),
+                                             CoinPortion, EpochIndex (..), ScriptVersion,
+                                             SoftforkRule (..), Timestamp (..),
+                                             ProtocolMagic (..))
+import           Pos.Core.Genesis.Types     (GenesisSpec (..), ProtocolConstants (..),
+                                             GenesisInitializer (..), GenesisHash (..),
+                                             GenesisData, genesisSpecFromData)
+import           Pos.Crypto.Hashing         (Hash, unsafeHash, AbstractHash (..))
+
+-- | The Configuration (probably from a yaml) and the system start time (which
+-- is not necessarily read from the Configuration).
+type HasConfiguration =
+    ( Given Configuration
+      -- System start time.
+    , Given Timestamp
+    , Given GenesisHash
+    , Given GenesisSpec
+    )
+
+-- | Come up with a HasConfiguration constraint using a Configuration.
+-- The Configuration record can be parsed from JSON or Yaml, and used to
+-- get a GenesisSpec, either from the file itself or from another file:
+-- the canonical JSON encoding of a mainnet genesis.
+--
+-- If the canonical JSON source is given, then it will be hashed and checked
+-- against the constant expected genesisDataDigest.
+--
+-- If the configuration gives a testnet genesis spec, then a start time must
+-- be provided, probably sourced from command line arguments.
+withConfiguration
+    :: ( MonadThrow m
+       , WithLogger m
+       , MonadIO m
+       , FromJSON (Either String) GenesisData
+       )
+    => Configuration
+    -> Maybe Timestamp
+    -- ^ Optional system start time.
+    --   It must be given when the genesis spec uses a testnet initializer.
+    -> (HasConfiguration => m r)
+    -> m r
+withConfiguration conf@Configuration{..} mSystemStart act = case ccGenesis of
+    -- If a 'GenesisData' source file is given, we check its hash against the
+    -- built-in constant, parse it, and we're done.
+    GCSrc fp -> do
+        !bytes <- liftIO $ BS.readFile fp
+
+        let digest :: Hash.Digest Hash.Blake2b_256
+            digest = Hash.hash bytes
```

Comment:
You could've used `Pos.Crypto.Hashing.hash` here.

Reply:
No, we need serialization to mess up things here!

Reply:
Sorry, I meant `hashRaw`.